### PR TITLE
#24307 Introducing composite storage chainable

### LINKED ISSUE #24307: Multi Metadata Provider/Chainable Metadata Provider
### Parent Issue

- [x] Store Assets in S3 

### User Story

Short Story Long: 
We built a cool FileStorage tool to store our assets File Metadata.  Initially, to access File Metadata, we tried using the `DataBaseStoragePersistenceAPIImpl` as DEFAULT  but this had a downside when running in at load in a cluster - each node in the cluster needed to do db hits to generate the metadata for every asset `SQL calls = (number of nodes * assets requested)`.  And when a node/container cycled, the new nodes needed to re-do all those sql queries.   This had the effect of swamping the DB and nodes were unable to process requests while they were trying to load metadata for the assets.

So we rolled back to the `FileSystemStoragePersistenceAPIImpl`, which mirrored the older behavior and stored the File Metadata on the NFS share.  This has the benefit that once one node has generated the metadata, the other nodes can pick it up with no load on the DB and that all metadata survives restarts.  But it has the downside that we are still reliant on the NFS share for metadata and assets and that adding the metadata file/folder takes up a number of fs inodes from the NFS share.

Where I would like to be is to have a combination of both - a chain where the Storage providers are checked in order.  In our case the `FileSystemStoragePersistenceAPIImpl` is checked first and if the metadata is not found, then the `DataBaseStoragePersistenceAPIImpl` is checked.  If the `DataBaseStoragePersistenceAPIImpl` finds the metadata then it returns it and it gets written to the NFS mount (or in the local filesystem, configurable) for future access and then returned as json to dotCMS.

The same would happen when we persist the metadata, we write it to the FileSystem first and then to the DB.

The point of a configurable chain is that 
1. By backing up the db we have a full snapshot of all asset metadata that will also perform well in a cluster, which is what we want.  We can then rely on the FS as a cache.
2. It would allow is to also write a `RedisStoragePersistenceAPIImpl` and configure it into the chain as well, after the NFS FS and before the DB or even use it to replace the NFS FS as the cluster persistence layer.
3. With these changes and confidence that we all our asset metadata is in the DB, we can start to store file assets by their sha256 hash, effectively de-duping them.
4. Once we are storing the assets by their sha256, we can start to store them in an s3 bucket 

### Acceptance Criteria

- Store file asset metadata in both the FS and the DB.
- Wipe out the fs metadata and be able to start a dotCMS and pull the metadata from the DB which would repopulate the FS metadata. 

### Proposed Objective

Cloud Engineering

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by jdotcms**
Implementing a StoragePersistenceAPI composite chainable;
- The composite encapsules two or more StoragePersistenceAPI
- When writing assets will write over all StoragePersistenceAPI collection
- When determining an state will query over all StoragePersistenceAPI collection, to see if any at least match
- When pulling assets will query all StoragePersistenceAPI collection, if any hit on the chain, will do backtracking over the upper layers of the chain, populating on async the storages with the recovery assets

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+
+        final List<Future<File>> futures = new ArrayList<>();
+        for(final StoragePersistenceAPI storage : this.storagePersistenceAPIList) {
+
+            final Future<File> localFileFuture = storage.pullFileAsync(groupName, path);
+            futures.add(localFileFuture);
+        }
+
+        // since we do not want to wait for all the futures to be completed, we do not which layer does not have the file, so we fire a new thread to call the sync method
+        DotConcurrentFactory.getInstance().getSubmitter().submit(()-> this.pullFile(groupName, path));
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java:296]: Should this be its own submitter group for greater configurability?

File: dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+        final List<Future<Object>> futures = new ArrayList<>();
+        for(final StoragePersistenceAPI storage : this.storagePersistenceAPIList) {
+
+            final Future<Object> localFileFuture = storage.pullObjectAsync(groupName, path, readerDelegate);
+            futures.add(localFileFuture);
+        }
+
+        // since we do not want to wait for all the futures to be completed, we do not which layer does not have the file, so we fire a new thread to call the sync method
+        DotConcurrentFactory.getInstance().getSubmitter().submit(()-> this.pullObject(groupName, path, readerDelegate));
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java:312]: Should this be its own submitter group for greater configurability?

File: dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java (hunk 1 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/storage/CompositeStoragePersistenceAPI.java:296]: ok sounds fair

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/storage/MetadataGeneratorImpl.java [hunk 3]
```diff
@@ -64,6 +65,7 @@ public Map<String, Serializable> tikaBasedMetadata(final File binary, final long
     public TreeMap<String, Serializable> standAloneMetadata(final File binary){
         final TreeMap<String, Serializable> metadataMap = new TreeMap<>(Comparator.naturalOrder());
         final String binaryName = binary.getName();
+        final String mimeType = Try.of(()->MimeTypeUtils.getMimeType(binary)).getOrElse(FileAsset.UNKNOWN_MIME_TYPE);
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/storage/MetadataGeneratorImpl.java:68]: Should we use `application/octet-stream` for our unknown mimeType?  Seems safer?

**Comment by wezell** [comment]
@jdotcms One important question - if this gets merged, does our current default behavior change - If it does change, then how does it change?

**Comment by jcastro-dotcms** [comment]
Re-opening this PR and merging the latest master into it introduced several Angular-related changes that were never part of the official code changes. Closing this PR and creating a new one.

**Pull request closed without merging**